vendor/github.com/prometheus/prometheus/storage/local/frankenstein.go

```diff
@@ -390,33 +387,43 @@ func (i *Ingester) flushAllUsers(immediate bool) {
 	i.userStateLock.Unlock()
 
 	for _, userID := range userIDs {
-		i.userStateLock.Lock()
-		userState, ok := i.userState[userID]
-		i.userStateLock.Unlock()
+		go i.flushUser(userID, immediate)
+	}
+}
 
-		// This should happen, right?
-		if !ok {
-			continue
-		}
+func (i *Ingester) flushUser(userID string, immediate bool) {
+	i.userStateLock.Lock()
+	userState, ok := i.userState[userID]
+	i.userStateLock.Unlock()
 
-		ctx := context.WithValue(context.Background(), UserIDContextKey, userID)
-		i.flushAllSeries(ctx, userState, immediate)
+	// This should happen, right?
+	if !ok {
+		return
+	}
 
-		// TODO: this is probably slow, and could be done in a better way.
-		i.userStateLock.Lock()
-		if userState.fpToSeries.length() == 0 {
-			delete(i.userState, userID)
-		}
-		i.userStateLock.Unlock()
+	ctx := user.WithID(context.Background(), userID)
+	i.flushAllSeries(ctx, userState, immediate)
+
+	// TODO: this is probably slow, and could be done in a better way.
+	i.userStateLock.Lock()
+	if userState.fpToSeries.length() == 0 {
+		delete(i.userState, userID)
 	}
+	i.userStateLock.Unlock()
 }
 
 func (i *Ingester) flushAllSeries(ctx context.Context, state *userState, immediate bool) {
+	var wg sync.WaitGroup
 	for pair := range state.fpToSeries.iter() {
-		if err := i.flushSeries(ctx, state, pair.fp, pair.series, immediate); err != nil {
-			log.Errorf("Failed to flush chunks for series: %v", err)
-		}
+		wg.Add(1)
+		go func() {
```

Comment:
Yes I wasn't sure either.  I could use a semaphore for customer instead of chunks, it seems a bit of a waste to start them all and have them immediately block.